jit: Introduce background compilation

qwe661234 opened the pull request:
Given the significant runtime compilation overhead associated with performing aggressive optimizations, we have implemented a background compilation mechanism to mitigate this issue. When the runtime profiler identifies a strong hotspot, it adds a T2C compilation request to the wait queue. A background thread, which continuously monitors this queue, triggers T2C to process the requests and notifies the main thread upon completion by updating a flag.

This mechanism defers the execution of T2C-generated machine code, leading to more frequent use of T1C-generated code. Despite this, the approach effectively minimizes runtime compilation delays by eliminating the need for the main thread to wait for T2C compilation to complete, thereby improving overall performance.

* The workflow of background compilation.
![image](https://github.com/sysprog21/rv32emu/assets/48278026/56e2bd62-deb1-4677-9a63-3bf283023080)

* The performance comparison with and without the background compialtion.
![image](https://github.com/sysprog21/rv32emu/assets/48278026/b6805c4d-0859-487f-b2d0-e3d1c4b42e08)

Close: #239

jserv reviewed (commented):
<!-- github-benchmark-action-comment(start): Benchmarks Summary -->
# Benchmarks

<details>

| Benchmark suite | Current: 51ce53be52b626176e58bde93664c0d86989377a | Previous: 6b462c03dc8c64863ae146a94fe7abffe4da19bb | Ratio |
|-|-|-|-|
| `Dhrystone` | `5.22` Average DMIPS over 10 runs | `4.88` Average DMIPS over 10 runs | `0.93` |
| `Coremark` | `0.004` Average iterations/sec over 10 runs | `0.004` Average iterations/sec over 10 runs | `1` |

</details>

This comment was automatically generated by [workflow](https://github.com/sysprog21/rv32emu/actions?query=workflow%3ABenchmark) using [github-action-benchmark](https://github.com/marketplace/actions/continuous-benchmark).
<!-- github-benchmark-action-comment(end): Benchmarks Summary -->

jserv commented on src/riscv.c:
```diff
@@ -184,6 +186,31 @@ IO_HANDLER_IMPL(byte, write_b, W)
 
+#if RV32_HAS(JIT)
+static pthread_t t2c_thread;
+void *t2c_routine(void *arg)
```
Should be `static`.

jserv commented on src/riscv_private.h:
```diff
@@ -132,8 +140,11 @@ struct riscv_internal {
 #if !RV32_HAS(JIT)
     block_map_t block_map; /**< basic block map */
 #else
+    bool exit /**< Determine the main thread is exit or not */;
```
Rename to `quit` in order to distinguish from the state of being exited from the execution or program being terminated. In this case, using `quit` is much clearer.

jserv commented on src/riscv.c:
```diff
@@ -269,6 +296,10 @@ riscv_t *rv_create(riscv_user_t rv_attr)
+    rv->exit = false;
+    pthread_mutex_init(&rv->wait_queue_lock, NULL);
+    INIT_LIST_HEAD(&rv->wait_queue);
+    pthread_create(&t2c_thread, NULL, t2c_routine, rv);
```
Add some comments here. E.g., activate the background compilation thread.

jserv commented on src/riscv.c:
```diff
@@ -184,6 +186,31 @@ IO_HANDLER_IMPL(byte, write_b, W)
 
+#if RV32_HAS(JIT) && RV32_HAS(T2C)
+static pthread_t t2c_thread;
+static void *t2c_routine(void *arg)
```
Rename to `t2c_runloop` to emphasize its behavior in background thread.

jserv commented:
I defer to @vacantron for confirmation.

jserv commented on src/riscv.c:
```diff
@@ -28,6 +28,8 @@
 #include "riscv_private.h"
 #include "utils.h"
 #if RV32_HAS(JIT)
+#include <pthread.h>
```
If T2C is not enabled, is pthread required?

jserv commented on src/riscv.c:
```diff
@@ -184,6 +186,31 @@ IO_HANDLER_IMPL(byte, write_b, W)
 #undef R
 #undef W
 
+#if RV32_HAS(JIT) && RV32_HAS(T2C)
```
We can centralize the dependency in `src/feature.h` to ensure T2C depends on JIT, eliminating the need to write `#if RV32_HAS(JIT) && RV32_HAS(T2C)`.

jserv commented on src/feature.h:
```diff
@@ -50,12 +50,11 @@
 #ifndef RV32_FEATURE_T2C
 #define RV32_FEATURE_T2C 0
 #endif
+#endif
```
Don't do that. Consider the following:
```c
#ifndef RV32_FEATURE_JIT
#define RV32_FEATURE_JIT 0
#endif

#ifndef RV32_FEATURE_T2C
#define RV32_FEATURE_T2C 0
#endif

/* T2C depends on JIT configuration */
#if !RV32_FEATURE_JIT
#undef RV32_FEATURE_T2C
#define RV32_FEATURE_T2C 0
#endif
```

jserv commented on src/riscv.c:
```diff
@@ -184,6 +187,31 @@ IO_HANDLER_IMPL(byte, write_b, W)
+                        (uint64_t) ((memory_t *) PRIV(rv)->mem)->mem_base);
+            free(entry);
+        }
+        /* Instead of writing while(rv->quit), placing the code here prevents
```
Does `volatile` help for such purpose?

qwe661234 replied:
Yes, it works.

qwe661234 committed:
Introduce background compilation

Given the significant runtime compilation overhead associated with
performing aggressive optimizations, we have implemented a background
compilation mechanism to mitigate this issue. When the runtime profiler
identifies a strong hotspot, it adds a T2C compilation request to the
wait queue. A background thread, which continuously monitors this queue,
triggers T2C to process the requests and notifies the main thread upon
completion by updating a flag.

jserv commented:
Finally, T2C is landed!
